lib/services/nodes-api-service.js

```diff
@@ -30,64 +32,136 @@ function nodeApiServiceFactory(
     }
 
     /**
-     * Find Enclosure nodes which enclose compute node
+     * Find target nodes that relate with the removing node
      * @param  {Object}     node
-     * @return {Promise}
+     * @param  {Object}     type
+     * @return {Promise}    array of target nodes
      */
-    NodeApiService.prototype._findEnclNodes = function(node) {
-        // Find the enclosure nodes who enclose this compute node
+    NodeApiService.prototype._findTargetNodes = function(node, type) {
         if (!_.has(node, 'relations')) {
             return Promise.resolve();
         }
 
-        var relation = _.find(node.relations, { relationType: 'enclosedBy' });
+        var relation = _.find(node.relations, { relationType: type });
         if (!relation || !_.has(relation, 'targets') ) {
             return Promise.resolve();
         }
 
-        return Promise.map(relation.targets, function (enclNodeId) {
-            return waterline.nodes.needByIdentifier(enclNodeId)
+        return Promise.map(relation.targets, function (targetNodeId) {
+            return waterline.nodes.needByIdentifier(targetNodeId)
             .catch(function (err) {
-                logger.warning("Error Getting Enclosure Node", { error: err });
+                logger.warning("Error Getting Downstream Node with type " + type,
+                    { error: err });
                 return;
             });
         });
     };
 
     /**
-     * Remove the relations between node and enclosure node, if enclosure node
-     * doesn't enclose any nodes, this enclosure node is also removed.
-     * @param  {Object}     node
-     * @param  {Array}      enclNodes
-     * @return {Promise}
+     * Remove the relations to node in the downstram node. If the relation field is
+     * empty after removing, delete this field.
+     * @param  {Object}     node node whose relation needs to be updated
+     * @param  {String}     type relation type that needs to be updated
+     * @param  {String}     targetId id in relation type that needs to be deleted
+     * @return {Promise}    node after removing relation
      */
-    NodeApiService.prototype._removeEnclRelations = function(node, enclNodes) {
-        if (!enclNodes) {
+    NodeApiService.prototype._removeRelation = function(node, type, targetId) {
+
+        if (!node) {
```

Comment:
Yeah, I see some drawbacks to both approaches:

If we enforce throwing an error, it means that we can never repair hung states through the API (e.g. if a target relation is deleted for some reason, we can now never successfully delete other nodes with a link relation to it).

If we don't throw an error, there could be cases where an intermittent connectivity failure to the database causes us to only partially delete relations when we really meant to delete all of them.

I'm not sure which drawback we want to avoid and which we want to accept. Thoughts?